EgammaAnalysis/src/PhotonHGCalIsoProducer.cc

```diff
@@ -0,0 +1,108 @@
+/*
+ * PhotonIsoProducer.cc
+ *
+ *  Created on: 13 Oct 2017
+ *      Author: jkiesele
+ */
+
+#include "EgammaTools/EgammaAnalysis/interface/PhotonHGCalIsoProducer.h"
+#include <stdexcept>
+
+PhotonHGCalIsoProducer::PhotonHGCalIsoProducer():dr_(0.15),mindr_(0),rechittools_(0),debug_(false){
+
+    allHitMap_ = new std::map<DetId, const HGCRecHit *>();
+    setNRings(5);
+}
+
+void PhotonHGCalIsoProducer::produceHGCalIso(const reco::Photon& photon){
+
+    if(!rechittools_)
+        throw std::runtime_error("PhotonHGCalIsoProducer::produceCaloIso: rechittools not set");
+
+    for(auto& r:isoringdeposits_)
+        r=0;
+
+    //this could be replaced by the hit map created by storeRecHits in PCAhelpers
+    std::vector<DetId>photonhits;
+    const std::vector<std::pair<DetId,float > > & photonhitmap= photon.superCluster()->seed()->hitsAndFractions();
+    for(const auto& h:photonhitmap)
+        photonhits.push_back(h.first);
+
+    for(const auto& hit: *allHitMap_) {
+
+        const GlobalPoint position = rechittools_->getPosition(hit.first);
+        float eta=rechittools_->getEta(position, 0);//assume vertex at z=0
+        float phi=rechittools_->getPhi(position);
+        float deltar=reco::deltaR(eta,phi,photon.eta(),photon.phi());
+
+        if(deltar>dr_ || deltar<mindr_) continue;
+
+        size_t layer=rechittools_->getLayerWithOffset(hit.first);
+        if(layer>=nlayers) continue;
+
+        const size_t& ring=ringasso_.at(layer);
+
+        //do not consider hits associated to the photon cluster
+        if(std::find(photonhits.begin(),photonhits.end(),hit.first)==photonhits.end()){
+            isoringdeposits_.at(ring)+=hit.second->energy();
+        }
+    }
+
+    //does this make sense here or should it be done later?
```

Comment:
@jkiesele I think leaving the raw energy might be safer than the ratio.  For example, the run 2 photon isolation cuts are only weakly pt-dependent.  Anyway the photon energy stored in that variable could be at some point a regression energy or some other thing that is not just sum(rechits).

Reply:
Fully agree with you.